ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Type.hs

```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{-
+ Block fetching consists of two protocols: submission of block ranges 'BlockFetchClientProtocol', and
+ streaming of blocks 'BlockFetchServerProtocol'.
+
+  client server
+   |       |
+   | \     |
+   |  \    | 'BlockFetchClientProtocol' (requesting a range)
+   |   \   |
+   |    \  |
+   |     \ |
+   |     / |
+   |    /  |
+   |\  / / | 'BlockFetchServerProtocol' (streaming blocks)
+   | \/ /  |
+   | /\/ / |
+   |  /\/  |
+   | / /\  |
+   |  /  \ |
+   | /   / |
+   |    /  |
+   |   / / |
+   |  / /  |
+   | / / / |
+   |  / /  |
+   | / /   |
+   |  /    |
+
+  Request submitted by @'BlockFetchClientProtocol'@ are recorded by the server
+  and supplied to the @'BlockFetchServerProtocol'@.  This way we achieve
+  protocol pipelining.
+
+  Both protocol never shift the agency: the 'BlockFetchClientProtocol' keep
+  agancy on the client side and the 'BlockFetchServerProtocol' keeps it on the
+  server side.
+
+  The receiver of @'BlockFetchClientProtocol'@ writes received ranges to
+  a queue which acumulates @'RangeStream' range@ elements.  The sender of
+  @'BlockFetchServerProtocol'@ is reading this queue and sending back blocks to
+  a @'BlockFetchServerReceiver'@.
+
+  Termination
+
+  If the @'BlockFetchClientSender'@ (of the @'BlockFetchClientProtoco'@) will
+  send @'MessageDone'@ the corresponding @'BlockFetchServerReciever'@ will
+  write @'End'@ to the queue which connects both protocols.  When the
+  @'BlockFetchServerSender'@ will encounter @'End'@, it will send
+  @'MessageServerDone'@ which will terminate the protocol.
+
+  Note: there is no mechanism which ensures that if the
+  @'BlockFetchServerSender'@ will send @'MessageServerDone'@ that the
+  corresponding @'BlockFetchClientProtocol@ will terminate.
+
+-}
+module Ouroboros.Network.Protocol.BlockFetch.Type where
+
+import Protocol.Core
+
+import Ouroboros.Network.Chain (Point)
+
+-- | Range of headers
+--
+data ChainRange header = ChainRange !(Point header) !(Point header)
```

Comment:
I think we should relax this so that the second one doesn't require a header hash. Sometimes we'll want to request up to a point for which we don't know the hash. A more useful representation is, I think:

```
data ChainRange header = ChainRange !(Point header) NumberOfBlocks
```

i.e. a start point and a (maximum) number of blocks from that point. If needed, we could also admit using slot count instead.